lsp/server/src/validateGraphql.ts

```diff
@@ -53,17 +68,31 @@ export async function validateGraphql(
 
 /**
  * Validate graphql diagnostic rules to a graph query, return empty list if the graphql string is invalid.
+ * @param producers The diagnostic producer to run.
  * @param graphql the graph code
  * @param graphqlDiagnosticProducers  the collection of graphql rules. 
  */
-export async function validateOneGraphQuery(textDocument: TextDocument, graphql: string): Promise<Diagnostic[]> {
+export async function validateOneGraphQuery(
+    producers: DiagnosticProducer<ASTNode>[],
+    textDocument: TextDocument, 
+    graphql: string
+): Promise<Diagnostic[]> {
   
     try {
         const graphqlAstNode = parse(graphql);
         const allResults = await Promise.all(
-            diagnosticProducers.map((producer) =>
-                producer.validateDocument(textDocument, graphqlAstNode)
-            )
+            producers.map((producer) => {
+                return producer.validateDocument(textDocument, graphqlAstNode)
+                 .then((diagnostics) => {
+                     const metaData: DiagnosticMetaData = {
+                         producerId: producer.getId()
+                     };
+                     diagnostics.forEach((diagnostic) => {
+                         diagnostic.data = metaData;
+                     });
```

Comment:
I understand why you made `DiagnosticMetaData`, that is, to take advantage of `Diagnostic` object's `data?` property, which unforunately is of type `any`.

But having a metadata that only has one property feels like an overkill. And the post processing of attaching such data on each of the diagnostic makes this argument stronger.

Can we devise of something else?
We can use [declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html). Extend `Diagnostic` interface in the `src/@types` folder
![image](https://github.com/user-attachments/assets/e0e0d540-c7b9-44be-a223-21db1cc57b1b)
Then you'll be able to add a property to `Diagnostic` and get rid of `DiagnosticMetaData`.

Reply:
let me give [declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html) a try, it will be great if property will be kept while the diagnostic is sent back and force between lsp client and server.

Reply:
tried, unfortunately the produceId is lost :( 
![image](https://github.com/user-attachments/assets/da65e915-764b-4aee-ade8-83cad100cbde)

I will use data to hold the produceId directly without the Metadata extra structure.  I over thought for future proof it.

Reply:
How funny. It worked on my machine.

Reply:
If this approach is also another overkill then I'd rather have your updated approach in simple string property in `DiagnosticProducer` 👍